controllers/argocd/util.go

```diff
@@ -1360,6 +1368,88 @@ func (r *ReconcileArgoCD) setManagedNamespaces(cr *argoproj.ArgoCD) error {
 	return nil
 }
 
+func (r *ReconcileArgoCD) setManagedSourceNamespaces(cr *argoproj.ArgoCD) error {
+	r.ManagedSourceNamespaces = make(map[string]string)
+	namespaces := &corev1.NamespaceList{}
+	listOption := client.MatchingLabels{
+		common.ArgoCDManagedByClusterArgoCDLabel: cr.Namespace,
+	}
+
+	// get the list of namespaces managed by the Argo CD instance
+	if err := r.Client.List(context.TODO(), namespaces, listOption); err != nil {
+		return err
+	}
+
+	for _, namespace := range namespaces.Items {
+		r.ManagedSourceNamespaces[namespace.Name] = ""
+	}
+
+	return nil
+}
+
+// removeUnmanagedSourceNamespaceResources cleansup resources from SourceNamespaces if namespace is not managed by argocd instance.
+// It also removes the managed-by-cluster-argocd label from the namespace
+func (r *ReconcileArgoCD) removeUnmanagedSourceNamespaceResources(cr *argoproj.ArgoCD) error {
+	if _, ok := r.ManagedSourceNamespaces[cr.Namespace]; !ok {
+		return nil
+	}
+
+	for _, ns := range r.ManagedSourceNamespaces {
+		managedNamespace := false
+		if cr.DeletionTimestamp == nil {
+			for _, namespace := range cr.Spec.SourceNamespaces {
+				if namespace == ns {
+					managedNamespace = true
+					break
+				}
+			}
+		}
```

Comment:
Do we need this check here? `r.ManagedSourceNamespaces` is populated by directly checking the cluster for which namespaces have the label assigned to the CR currently being reconciled 
This check seems unnecessary to me

Reply:
Ah sorry  I think i see what you're doing here - trying to synchronize `r.ManagedSourceNamespaces` with `cr.spec.SourceNamespaces`
but I think there might be a better way to handle that 


Here we will be running this whole loop each time the operator reconciles, even when no changes to `cr.spec.SourceNamespaces` has been made
instead maybe we could add a predicate here that watches for changes to `cr.spec.SourceNamespaces`. We can compare the lists in the `oldCR` and `newCR` by checking if any namespace is present in `oldCR` but not in `newCR`
for those namespaces we can invole `removeUnmanagedSourceNamespaceResources`
that way we only invoke this function when some actual change has been detected to the list of managed source namespaces

let me know what you think

Reply:
I think if we add a predicate to the check for changes in SourceNamespaces, we might not be able to prioritize ManagedBy label over MangedByClusterArgocd label. 

For example, consider the scenario where the namespace was added as SourceNamespace and then the ManagedBy label was added to the namespace. In this case, we would reconcile resources for the namespace only for ManagedBy label and not for MangedByClusterArgocd label.

I think then the question would be, do we need to prioritize ManagedBy label over ManagedByClusterArgocd label as we are reconciling ManagedBy label resources first?

Lemme know WDYT..

Reply:
@ishitasequeira as another small optimization maybe we can rewrite this code to use only one loop instead of a nested one

```
for _, ns := rnage cr.Spec.SourceNamespaces {
   managedNamespace := false 
   if cr.DeletionTimestamp == nil {
     if _, ok := r.ManagedSoruceNamespaces[ns]; ok {
       managedNamesapce = true
       break
       }
     }
   }
```

let me know what you think

Reply:
In the double for loop, what we are trying to do is, we are trying to find out which of the `r.ManagedSoruceNamespaces` is not part of `cr.Spec.SourceNamespaces` so as to perform the cleanup in that namespace. The for loop you have suggested would not help in achieving the same.

Reply:
yeah sorry you're right 
maybe `sourceNamespaces` could also be a map to be honest

Reply:
`sourceNamespaces` is part of the  ArgoCD CR. IMO, it would not make sense to ask users to send a map when they just need to send a list of namespaces.

Reply:
I meant we can make a map for it before accessing it in the backend logic